### PR TITLE
Integration data type test for new table format

### PR DESCRIPTION
**Pull request by sfc-gh-alhuang**
1. Created data type ITs for Iceberg table and verified through local VM. This test should be disabled until the sever side change is release.
2. Bug fix for `FileColumnProperties`. The Iceberg mode will fill in default value for corresponding type now if all values in the column are null.
3. Created `AbstractDataTypeTest.testIcebergIngestAndQuery` for multi row testing.

### CODE REVIEW

File: src/test/java/net/snowflake/ingest/streaming/internal/datatypes/AbstractDataTypeTest.java [hunk 1]
```diff
@@ -112,6 +122,21 @@ protected String createTable(String dataType) throws SQLException {
...
+
+  protected String createIcebergTable(String dataType) throws SQLException {
+    String tableName = getRandomIdentifier();
+    conn.createStatement()
+        .execute(
+            String.format(
+                "create or replace iceberg table %s (%s string, %s %s) "
+                    + "catalog = 'SNOWFLAKE' "
+                    + "external_volume = 'streaming_ingest' "
+                    + "base_location = 'all_type_col';",
```
Comment (sfc-gh-hmadan) [inline src/test/java/net/snowflake/ingest/streaming/internal/datatypes/AbstractDataTypeTest.java:137]: 1. lets use a different base_location for each table.
2. Create a JIRA for how to cleanup all of the data that ITs upload, from the external volume.

File: src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeTest.java [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+import net.snowflake.ingest.streaming.SnowflakeStreamingIngestChannel;
+import net.snowflake.ingest.utils.ErrorCode;
+import net.snowflake.ingest.utils.SFException;
+import org.apache.commons.lang3.StringUtils;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Ignore;
+import org.junit.Test;
+
+@Ignore("This test can be enabled after server side Iceberg EP support is released")
```
Comment (sfc-gh-hmadan) [inline src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeTest.java:25]: Add a backlog JIRA to track adding a different github merge gate for iceberg ITs, we'll then have one merge gate for FDN ITs and another for iceberg ITs.

File: src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeTest.java [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+  public void before() throws Exception {
+    super.before(true);
+  }
+
+  @Test
+  public void testBoolean() throws Exception {
+    testIcebergIngestion("boolean", true, new BooleanProvider());
+    testIcebergIngestion("boolean", false, new BooleanProvider());
+    testIcebergIngestion("boolean", 1, true, new BooleanProvider());
+    testIcebergIngestion("boolean", "false", false, new BooleanProvider());
```
Comment (sfc-gh-hmadan) [inline src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeTest.java:40]: lets also add 0, "true", "1", "0"

File: src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeTest.java [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+  private final ObjectMapper objectMapper = new ObjectMapper();
+
+  @Before
+  public void before() throws Exception {
+    super.before(true);
+  }
+
+  @Test
+  public void testBoolean() throws Exception {
+    testIcebergIngestion("boolean", true, new BooleanProvider());
```
Comment (sfc-gh-hmadan) [inline src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeTest.java:37]: nit: whenever you have time in the next few weeks, lets introduce a class called ITTestcaseBuilder that has methods withDataType, withInputValue, withExpectedValue, withVerificationQuery.

THe provider business is funky as it only supports doing exact value verification with a `where value = {expectedValue}`. I'd much rather write the sql filter directly in the testcase instead of have code generate it, for flexibility.

File: src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeTest.java [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+    testIcebergIngestion("boolean", false, new BooleanProvider());
+    testIcebergIngestion("boolean", 1, true, new BooleanProvider());
+    testIcebergIngestion("boolean", "false", false, new BooleanProvider());
+
+    SFException ex =
+        Assert.assertThrows(
```
Comment (sfc-gh-hmadan) [inline src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeTest.java:43]: pl use fluent assertions (assertj) as they (a) have test code that reads like a story, and (b) generate very well formatted error messages. Assert.assert*() has the most useless error messages when things fail unless every single call to the method has an associated string message that details the current state of whatever was being asserted on.

For instance:
assertThat(ex).isNotNull();
assertThat(ex.getVendorCode).isEqualTo(ErrorCode.INVALID_FORMAT_ROW.getMessageCode());
assertThatCode(() -> lambda).doesNotThrow();
assertThatThrownBy(() -> lambda).hasCause();

For this you'll have to add the assertj library (I added it once and it failed in one of the mergegates, did not have the time to find and fix what happened!). Feel free to remove the org.hamcrest import from this project.

File: src/test/java/net/snowflake/ingest/streaming/internal/datatypes/AbstractDataTypeTest.java [hunk 6]
```diff
@@ -112,6 +122,21 @@ protected String createTable(String dataType) throws SQLException {
             String.format(
                 "create or replace table %s (%s string, %s %s)",
                 tableName, SOURCE_COLUMN_NAME, VALUE_COLUMN_NAME, dataType));
+
+    return tableName;
+  }
+
+  protected String createIcebergTable(String dataType) throws SQLException {
```
Comment (sfc-gh-hmadan) [inline src/test/java/net/snowflake/ingest/streaming/internal/datatypes/AbstractDataTypeTest.java:129]: also add an argument for `boolean nullable` and we can make sure the sdk does proper nullability checking for columns marked non-nullable.

File: src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeTest.java [hunk 7]
```diff
@@ -0,0 +1,441 @@
...
+            SFException.class,
+            () -> testIcebergIngestion("boolean", new Object(), true, new BooleanProvider()));
+    Assert.assertEquals(ErrorCode.INVALID_FORMAT_ROW.getMessageCode(), ex.getVendorCode());
```
Comment (sfc-gh-hmadan) [inline src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeTest.java:46]: (dupe) lets add nullability tests too - sending in null should work for every column, but when its a non-null column then the sdk should throw.

File: src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeTest.java [hunk 8]
```diff
@@ -0,0 +1,441 @@
...
+    testIcebergIngestion("int", 0.5f, 1, new IntProvider());
+    testIcebergIngestion("int", "100.4", 100, new IntProvider());
+    testIcebergIngestion("int", new BigDecimal("1000000.09"), 1000000, new IntProvider());
+    testIcebergIngestion("int", Integer.MAX_VALUE, new IntProvider());
+    testIcebergIngestion("int", Integer.MIN_VALUE, new IntProvider());
+
+    SFException ex =
+        Assert.assertThrows(
+            SFException.class,
+            () -> testIcebergIngestion("int", Long.MAX_VALUE, new LongProvider()));
```
Comment (sfc-gh-hmadan) [inline src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeTest.java:62]: nit: `(Long)Integer.MAX_VALUE + 1` instead of long.maxvalue?

File: src/test/java/net/snowflake/ingest/streaming/internal/datatypes/AbstractDataTypeTest.java (hunk 1 above)
Comment (sfc-gh-alhuang) [reply src/test/java/net/snowflake/ingest/streaming/internal/datatypes/AbstractDataTypeTest.java]: 1. Set up different base_location for each tests.
2. [JIRA](https://snowflakecomputing.atlassian.net/browse/SNOW-1727545) created.

File: src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeTest.java (hunk 2 above)
Comment (sfc-gh-alhuang) [reply src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeTest.java]: [jira](https://snowflakecomputing.atlassian.net/browse/SNOW-1727546) created.

File: src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeTest.java (hunk 3 above)
Comment (sfc-gh-alhuang) [reply src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeTest.java]: Added.

File: src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeTest.java (hunk 4 above)
Comment (sfc-gh-alhuang) [reply src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeTest.java]: Added to [jira](https://snowflakecomputing.atlassian.net/browse/SNOW-1727532)

File: src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeTest.java (hunk 5 above)
Comment (sfc-gh-alhuang) [reply src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeTest.java]: Switched to assertj library.

File: src/test/java/net/snowflake/ingest/streaming/internal/datatypes/AbstractDataTypeTest.java (hunk 6 above)
Comment (sfc-gh-alhuang) [reply src/test/java/net/snowflake/ingest/streaming/internal/datatypes/AbstractDataTypeTest.java]: Added nullable test.

File: src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeTest.java (hunk 8 above)
Comment (sfc-gh-alhuang) [reply src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeTest.java]: Modified.

File: src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeTest.java (hunk 7 above)
Comment (sfc-gh-alhuang) [reply src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeTest.java]: Added nullable test for non structured data type. Will fix structured data type null count in next PR.

File: src/main/java/net/snowflake/ingest/streaming/internal/RowBufferStats.java [hunk 9]
```diff
@@ -37,19 +38,58 @@ class RowBufferStats {
   private final String collationDefinitionString;
   /** Display name is required for the registration endpoint */
   private final String columnDisplayName;
+  /** Whether the column is an Iceberg column */
+  private final boolean isIcebergColumn;
+  /** Primitive type of the column, only used for Iceberg columns */
+  private final PrimitiveType primitiveType;
```
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/RowBufferStats.java:44]: pl update the PR description to explain there's a bug fix in this PR too.

File: src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java [hunk 10]
```diff
@@ -649,7 +649,7 @@ static EpInfo buildEpInfoFromStats(
     EpInfo epInfo = new EpInfo(rowCount, new HashMap<>());
```
Comment (sfc-gh-hmadan) [inline src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java:649]: pl cleanup the setDefaultValues from the whole callstack and from the parameter provider.

Comment (sfc-gh-hmadan) [reply src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java:649]: ignore this, the stacked PR I'm sending actually uses this param.

File: src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeIT.java [hunk 11]
```diff
@@ -0,0 +1,716 @@
...
+import net.snowflake.ingest.utils.ErrorCode;
+import net.snowflake.ingest.utils.SFException;
+import org.apache.commons.lang3.StringUtils;
+import org.assertj.core.api.Assertions;
+import org.junit.Before;
+import org.junit.Ignore;
+import org.junit.Test;
+
+@Ignore("This test can be enabled after server side Iceberg EP support is released")
+public class IcebergDataTypeIT extends AbstractDataTypeTest {
```
Comment (sfc-gh-hmadan) [inline src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeIT.java:27]: why not split this into multiple classes?

File: src/main/java/net/snowflake/ingest/streaming/internal/RowBufferStats.java [hunk 12]
```diff
@@ -37,19 +38,58 @@ class RowBufferStats {
...
+        null /* primitiveType */);
+  }
+
+  RowBufferStats(
+      String columnDisplayName,
+      String collationDefinitionString,
+      int ordinal,
+      Integer fieldId,
+      boolean isIcebergColumn,
+      PrimitiveType primitiveType) {
```
[the remark this answers is not in the document]
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/RowBufferStats.java]: Thanks for the fix!

File: src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeIT.java (hunk 11 above)
Comment (sfc-gh-alhuang) [reply src/test/java/net/snowflake/ingest/streaming/internal/datatypes/IcebergDataTypeIT.java]: Split into multiple test classes as FDN data type IT did.

**Comment by sfc-gh-alhuang** [comment]
@sfc-gh-hmadan Confirmed that the string min/max values for Iceberg table EP info also apply the 32 bytes round down/up truncated as FDN did.

File: src/test/java/net/snowflake/ingest/streaming/internal/datatypes/AbstractDataTypeTest.java (hunk 6 above)
Comment (sfc-gh-hmadan) [reply src/test/java/net/snowflake/ingest/streaming/internal/datatypes/AbstractDataTypeTest.java:130]: I see nullable was removed, why so?

Comment (sfc-gh-alhuang) [reply src/test/java/net/snowflake/ingest/streaming/internal/datatypes/AbstractDataTypeTest.java:130]: Removed and use `<datatype> not null` in datatype testing. By doing this we can set nullable on any sub-columns on structured data type.
